src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp

```diff
@@ -0,0 +1,840 @@
+                ap_pdf[3] / (2.0f * pi);
+
+            if (std::abs(wi.y) > 0)
+                bsdf_f /= std::abs(wi.y);
+
+            sample.set_to_scattering(ScatteringMode::Glossy, bsdf_pdf);
+
+            sample.m_value.m_glossy = bsdf_f;
+            sample.m_value.m_beauty = bsdf_f;
+
+            sample.m_incoming = Dual3f(sample.m_shading_basis.transform_to_parent(wi));
+
+        }
+
+        float evaluate(
+            const void*                 data,
+            const bool                  adjoint,
+            const bool                  cosine_mult,
+            const Vector3f&             geometric_normal,
+            const Basis3f&              shading_basis,
+            const Vector3f&             outgoing,
+            const Vector3f&             incoming,
+            const int                   modes,
+            DirectShadingComponents&    value) const override
+        {
+            if (!ScatteringMode::has_glossy(modes))
+                return 0.0f;
+
+            // Compute the BRDF value.
+            const InputValues* values = static_cast<const InputValues*>(data);
+
+            const Vector3f wo = shading_basis.transform_to_local(outgoing);
+            const Vector3f wi = shading_basis.transform_to_local(incoming);
+
+            // Compute outgoing terms
+            float sin_theta_o = wo.x;
+            float cos_theta_o = std::sqrt(std::max(0.0f, 1.0f - sin_theta_o * sin_theta_o));
+            float phi_o = std::atan2(wo.y, wo.z);
+            float gamma_o = std::asin(clamp(values->m_precomputed.m_h, -1.0f, 1.0f));
+
+            // Compute incoming terms
+            float sin_theta_i = wi.x;
+            float cos_theta_i = std::sqrt(std::max(0.0f, 1.0f - sin_theta_i * sin_theta_i));
+            float phi_i = std::atan2(wi.y, wi.z);
+
+            // Compute terms for the refracted ray
+            float sin_theta_t = sin_theta_o / values->m_eta;
+            float cos_theta_t = std::sqrt(std::max(0.0f, 1.0f - sin_theta_t * sin_theta_t));
+
+            // Compute the modified refraction coefficient
+            float eta_p = std::sqrt(values->m_eta * values->m_eta - sin_theta_o * sin_theta_o) / cos_theta_o;
+            float sin_gamma_t = values->m_precomputed.m_h / eta_p;
+            float cos_gamma_t = std::sqrt(std::max(0.0f, 1.0f - sin_gamma_t * sin_gamma_t));
+            float gamma_t = std::asin(clamp(sin_gamma_t, -1.0f, 1.0f));
+            float theta_o = std::asin(clamp(sin_theta_o, -1.0f, 1.0f));
+
+            // Deviation of outgoing angle due to cuticular scales
+            float theta_o_R = theta_o - 2.0f * deg_to_rad(values->m_alpha);
+            float theta_o_TT = theta_o + deg_to_rad(values->m_alpha);
+            float theta_o_TRT = theta_o + 4.0f * deg_to_rad(values->m_alpha);
+
+            // Compute Attenuation for the single path through the hair cylinder
+            Spectrum T = exp(-values->m_precomputed.m_sigma_a * ((2.f * cos_gamma_t) / cos_theta_t));
```

Comment:
`std::exp()`

Reply:
This is the exp() function from dynamicspectrum.h, not the one from the standard library.